### PR TITLE
Handle templates for postmark

### PR DESCRIPTION
**Pull request by KamilLelonek**
[Postmark](https://postmarkapp.com/) allows using [predefined templates](http://developer.postmarkapp.com/developer-api-templates.html) instead of sending `HTML` or `TEXT` body.

To do that, we need to provide 2 things:

- `TemplateId` - which is the ID of template to use when sending an email defined in Postmark account
- `TemplateModel` - which presents the model to be applied to the specified template to generate `HtmlBody`, `TextBody`, and `Subject`

This PR is supposed to give users a possibility to send email using these templates instead of defining custom templates from the code.

### CODE REVIEW

File: lib/swoosh/adapters/postmark.ex [hunk 1]
```diff
@@ -67,26 +75,53 @@ defmodule Swoosh.Adapters.Postmark do
...
+  defp prepare_html(body, %Email{html_body: html_body}), do: Map.put(body, "HtmlBody", html_body)
+
+  # example custom vars
+  #
+  # %{
+  #   "template_id"    => 123,
+  #   "template_model" => %{"name": 1, "company": 2}
+  # }
+  defp prepare_custom_vars(body, %Email{provider_options: provider_options}),
+    do: Enum.reduce(provider_options, body, &put_in_body/2)
```
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: Feels dangerous to reduce the whole `provider_options` into the body of the request.

Feels like this would be better handled by a specific `prepare_template/2` function that would match exactly on these two keys and insert them into the body.

File: lib/swoosh/adapters/postmark.ex [hunk 2]
```diff
@@ -67,26 +75,53 @@ defmodule Swoosh.Adapters.Postmark do
...
+  defp prepare_custom_vars(body, _email), do: body
+
+  defp put_in_body({key, val}, body_acc) do
+    key = key |> to_string() |> Macro.camelize()
```
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: See comment above, but in any case we shouldn't use the `Macro.camelize/1` function

File: test/integration/adapters/postmark_test.exs [hunk 3]
```diff
@@ -5,23 +5,42 @@ defmodule Swoosh.Integration.Adapters.PostmarkTest do
 
   @moduletag integration: true
 
-  setup_all do
-    config = [api_key: System.get_env("POSTMARK_API_KEY"), domain: System.get_env("POSTMARK_DOMAIN")]
-    {:ok, config: config}
-  end
+  @config [
```
Comment (stevedomin) [inline test/integration/adapters/postmark_test.exs]: This changes the behaviour as it will be grabbing the environment variables at compile-time rather than runtime, that can/will cause problem in some environments.

File: test/integration/adapters/postmark_test.exs [hunk 4]
```diff
@@ -5,23 +5,42 @@ defmodule Swoosh.Integration.Adapters.PostmarkTest do
...
+    api_key: System.get_env("POSTMARK_API_KEY"),
+    domain:  System.get_env("POSTMARK_DOMAIN"),
+  ]
+
+  @base_email new
```
Comment (stevedomin) [inline test/integration/adapters/postmark_test.exs]: See comment about compile-time/runtime above. This would be nicer as a function.

File: test/swoosh/adapters/postmark_test.exs [hunk 5]
```diff
@@ -15,9 +16,11 @@ defmodule Swoosh.Adapters.PostmarkTest do
   """
 
   setup_all do
-    bypass = Bypass.open
-    config = [base_url: "http://localhost:#{bypass.port}",
-              api_key: "jarvis"]
+    bypass = Bypass.open()
```
Comment (stevedomin) [inline test/swoosh/adapters/postmark_test.exs]: Style change, should be in a separate PR and across all adapters.

File: test/swoosh/adapters/postmark_test.exs [hunk 6]
```diff
@@ -15,9 +16,11 @@ defmodule Swoosh.Adapters.PostmarkTest do
...
+    config = [
```
Comment (stevedomin) [inline test/swoosh/adapters/postmark_test.exs]: Style change, should be in a separate PR and across all adapters.

File: test/swoosh/adapters/postmark_test.exs [hunk 7]
```diff
@@ -31,14 +34,17 @@ defmodule Swoosh.Adapters.PostmarkTest do
...
+        "To"       => "tony.stark@example.com",
+        "From"     => "steve.rogers@example.com",
+        "HtmlBody" => "<h1>Hello</h1>"
+      }
+
       assert body_params == conn.body_params
-      assert "/email" == conn.request_path
-      assert "POST" == conn.method
+      assert "/email"    == conn.request_path
+      assert "POST"      == conn.method
```
Comment (stevedomin) [inline test/swoosh/adapters/postmark_test.exs]: All the above: Style change, should be in a separate PR and across all adapters.

File: test/swoosh/adapters/postmark_test.exs [hunk 8]
```diff
@@ -63,51 +69,95 @@ defmodule Swoosh.Adapters.PostmarkTest do
...
+        "ReplyTo"  => "iron.stark@example.com",
+        "TextBody" => "Hello",
+        "HtmlBody" => "<h1>Hello</h1>"
+      }
 
       assert body_params == conn.body_params
-      assert "/email" == conn.request_path
-      assert "POST" == conn.method
+      assert "/email"    == conn.request_path
+      assert "POST"      == conn.method
```
Comment (stevedomin) [inline test/swoosh/adapters/postmark_test.exs]: All the above: Style change, should be in a separate PR and across all adapters.

**Comment by KamilLelonek** [comment]
Changes were made accordingly to your remarks.

File: lib/swoosh/adapters/postmark.ex (hunk 1 above)
Comment (KamilLelonek) [inline lib/swoosh/adapters/postmark.ex]: Isn't it done https://github.com/swoosh/swoosh/blob/master/lib/swoosh/adapters/mailgun.ex#L76-L80 here the same?

File: lib/swoosh/adapters/postmark.ex [hunk 9]
```diff
@@ -8,8 +8,9 @@ defmodule Swoosh.Adapters.Postmark do
 
       # config/config.exs
       config :sample, Sample.Mailer,
-        adapter: Swoosh.Adapters.Postmark,
-        api_key: "my-api-key"
+        adapter:  Swoosh.Adapters.Postmark,
+        api_key:  "my-api-key",
+        template: true # optional
```
Comment (KamilLelonek) [inline lib/swoosh/adapters/postmark.ex]: Not really, we should pick a different URL based on that: https://github.com/KamilLelonek/swoosh/blob/handle-templates-for-postmark/lib/swoosh/adapters/postmark.ex#L57-L61

File: lib/swoosh/adapters/postmark.ex (hunk 1 above)
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: Not exactly, in the Mailgun adapter only the `custom_vars` key inside the `provider_options` is reduce on the template.

File: lib/swoosh/adapters/postmark.ex (hunk 9 above)
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: What prevents us from checking if `provider_config["template_id"]` is set?

Even if you want to go with the separate `template` config I feel like this should go in the `provider_options` rather than being global.

File: lib/swoosh/adapters/postmark.ex (hunk 1 above)
Comment (KamilLelonek) [inline lib/swoosh/adapters/postmark.ex]: Anyway, I changed that.

File: lib/swoosh/adapters/postmark.ex (hunk 9 above)
Comment (KamilLelonek) [inline lib/swoosh/adapters/postmark.ex]: Because constructing endpoint happens earlier - during initialization, not during email delivery. We shouldn't modify endpoint later on but do that when defining an adapter https://github.com/KamilLelonek/swoosh/blob/79e0b7bef1692a9f7ebc92d34a663f16db2d4e1d/lib/swoosh/adapters/postmark.ex#L58-L59

Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: The construction of the endpoint happens when you call `deliver/2`, so it happens every time you send an email.

I really believe this should be attached to the email rather than a config on the adapter.

**Comment by stevedomin** [comment]
Still a couple of style changes that are cluttering the diff, once these are reverted this will be almost ready to be merged

File: lib/swoosh/adapters/postmark.ex (hunk 9 above)
Comment (KamilLelonek) [inline lib/swoosh/adapters/postmark.ex]: How would you prepare config then? Based on `%Email{}`?

Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: Yes, you can pass the email down to the `api_endpoint/1` function:

```
defp api_endpoint(%Email{provider_options: %{template: true}}), do: @api_endpoint <> "/withTemplate"
defp api_endpoint(_email), do: @api_endpoint
```

File: lib/swoosh/adapters/postmark.ex [hunk 10]
```diff
@@ -26,30 +27,35 @@ defmodule Swoosh.Adapters.Postmark do
 
   def deliver(%Email{} = email, config \\ []) do
     headers = prepare_headers(config)
-    params = email |> prepare_body |> Poison.encode!
```
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: I would keep this as is.

File: lib/swoosh/adapters/postmark.ex [hunk 11]
```diff
@@ -26,8 +26,9 @@ defmodule Swoosh.Adapters.Postmark do
...
-    url = [base_url(config), @api_endpoint]
+    email   = prepare_body(email)
+    url     = prepare_url(config, email)
```
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: And use the email (`%Email{}`) here.

File: lib/swoosh/adapters/postmark.ex [hunk 12]
```diff
@@ -39,15 +40,24 @@ defmodule Swoosh.Adapters.Postmark do
...
      {"Accept", "application/json"}]
   end
 
+  defp prepare_url(config, email),
+    do: [base_url(config), api_endpoint(email)]
+
+  defp base_url(config),
+    do: config[:base_url] || @base_url
+
+  defp api_endpoint(%{"TemplateModel" => _, "TemplateId" => _}),
```
Comment (stevedomin) [inline lib/swoosh/adapters/postmark.ex]: Which means you'd get the `%Email{}` struct here.

That way it's consistent with the rest of the file.

Comment (KamilLelonek) [inline lib/swoosh/adapters/postmark.ex]: Done.
